### PR TITLE
[5.2] Changed the Validator::validateUrl method to use the Symfony regex

### PR DESCRIPTION
**Pull request by scrothers**
This is in response to #10149, with this the PR from @franzliedke should pass based on @taylorotwell suggestion to use the Symphony regex for URL validation.

This is my first PR to Laravel, so I'm not sure I did this 100% correctly... apologies if it's wrong.

### CODE REVIEW

**Comment by crynobone** [comment]
@scrothers would be useful if you include the "failing" tests from #10149 to ensure that this solve it.

**Comment by scrothers** [comment]
Good point, but I made one change to the tests that didn't make sense.

The url 'http://laravel.com/?' in the original issue, is listed as "invalid" however, that is most definitely a valid url. So I marked that as a valid url.

**Comment by Sladewill** [comment]
Did you remove it from the invalid list because it looks like its still in there?

**Comment by scrothers** [comment]
You're looking at 'http://laravel.com?' not 'http://laravel.com/?', notice the last slash in particular.

File: src/Illuminate/Validation/Validator.php [hunk 1]
```diff
@@ -1200,7 +1201,13 @@ protected function validateEmail($attribute, $value)
...
...ircs|iris|iris.beep|iris.lwz|iris.xpc|iris.xpcs|itms|jabber|jar|jms|keyparc|lastfm|ldap|ldaps|magnet|mailserver|mailto|maps|market|message|mid|mms|modem|ms-help|ms-settings|ms-settings-airplanemode|ms-settings-bluetooth|ms-settings-camera|ms-settings-cellular|ms-settings-cloudstorage|ms-settings-emailandaccounts|ms-settings-language|ms-settings-location|ms-settings-lock|ms-settings-nfctransactions|ms-settings-notifications|ms-settings-power|ms-settings-privacy|ms-settings-proximity|ms-settings-screenrotation|ms-settings-wifi|ms-settings-workplace|msnim|msrp|msrps|mtqp|mumble|mupdate|mvn|news|nfs|ni|nih|nntp|notes|oid|opaquelocktoken|pack|palm|paparazzi|pkcs11|platform|pop|pres|prospero|proxy|psyc|query|redis|rediss|reload|res|resource|rmi|rsync|rtmfp|rtmp|rtsp|rtsps|rtspu|secondlife|service|session|sftp|sgn|shttp|sieve|sip|sips|skype|smb|sms|smtp|snews|snmp|soap.beep|soap.beeps|soldat|spotify|ssh|steam|stun|stuns|submit|svn|tag|teamspeak|tel|teliaeid|telnet|tftp|things|thismessage|tip|tn3270|turn|turns|tv|udp|unreal|urn|ut2004|vemmi|ventrilo|videotex|view-source|wais|webcal|ws|wss|wtai|wyciwyg|xcon|xcon-userid|xfire|xmlrpc.beep|xmlrpc.beeps|xmpp|xri|ymsgr|z39.50|z39.50r|z39.50s?)';
```
Comment (scrothers) [inline src/Illuminate/Validation/Validator.php]: This was derived from the [iana uri scheme list](http://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml) as suggested by @borys-p in #10149

**Comment by Sladewill** [comment]
Ah yeah sorry didn't see there wasn't a slash in it.

**Comment by franzliedke** [comment]
I think the idea was to copy the regex, not include the Symfony library, but obviously there are also benefits to that approach. :)

File: composer.json [hunk 2]
```diff
@@ -38,6 +38,7 @@
         "symfony/process": "3.0.*",
         "symfony/routing": "3.0.*",
         "symfony/translation": "3.0.*",
+        "symfony/validator": "3.0.*",
```
Comment (GrahamCampbell) [inline composer.json]: you need to update BOTH composer.json files pleae

**Comment by scrothers** [comment]
@GrahamCampbell I added it to the other composer.json as well. Should build fine in Travis in a few minutes.

**Comment by taylorotwell** [comment]
I don't think we should pull in the component. I think we should just lift the regex.

**Comment by scrothers** [comment]
@taylorotwell Done, I removed the inclusion of the Symfony library and also that sprintf by just including the URIs into the regex itself. I considered adding the regex to the class as a constant, like in Symfony, but I felt that was only going to serve to make the code messy, so I included the regex into the method. Open to suggestion there, that decision was made based on code organization only and no technical reason.

File: src/Illuminate/Validation/Validator.php [hunk 3]
```diff
@@ -1200,7 +1200,27 @@ protected function validateEmail($attribute, $value)
      */
     protected function validateUrl($attribute, $value)
     {
-        return filter_var($value, FILTER_VALIDATE_URL) !== false;
+        $pattern = '~^
```
Comment (GrahamCampbell) [inline src/Illuminate/Validation/Validator.php]: we need to attribute where this regex came from - symfony still owns the rights for it

File: src/Illuminate/Validation/Validator.php [hunk 4]
```diff
@@ -1200,7 +1200,27 @@ protected function validateEmail($attribute, $value)
...
+                \]  # a IPv6 address
+            )
+            (:[0-9]+)?                              # a port (optional)
+            (/?|/\S+)                               # a /, nothing or a / with something
+        $~ixu';
+
+        if (preg_match($pattern, $value) == 1) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Validation/Validator.php]: just `return preg_match($pattern, $value) == 1);`

Comment (GrahamCampbell) [inline src/Illuminate/Validation/Validator.php]: no if statement needed

File: tests/Validation/ValidationValidatorTest.php [hunk 5]
```diff
@@ -963,16 +963,281 @@ public function testValidateEmail()
...
     }
 
-    public function testValidateUrl()
+    /**
+     * @dataProvider validUrls
+     */
+    public function testValidateUrlWithValidUrls($validUrl)
     {
         $trans = $this->getRealTranslator();
-        $v = new Validator($trans, ['x' => 'aslsdlks'], ['x' => 'Url']);
```
Comment (GrahamCampbell) [inline tests/Validation/ValidationValidatorTest.php]: why was this deleted?

Comment (GrahamCampbell) [inline tests/Validation/ValidationValidatorTest.php]: oh, right, i see, at the bottom

File: src/Illuminate/Validation/Validator.php (hunk 4 above)
Comment (scrothers) [inline src/Illuminate/Validation/Validator.php]: Good point, I didn't think that would work for some reason because of the int return instead of a bool return on preg_match. Tested it and it works fine.

File: src/Illuminate/Validation/Validator.php (hunk 3 above)
Comment (scrothers) [inline src/Illuminate/Validation/Validator.php]: Done, good point.

File: src/Illuminate/Validation/Validator.php [hunk 6]
```diff
@@ -1200,7 +1200,27 @@ protected function validateEmail($attribute, $value)
      */
     protected function validateUrl($attribute, $value)
     {
-        return filter_var($value, FILTER_VALIDATE_URL) !== false;
+        /*
+         * This pattern is derived from Symfony\Component\Validator\Constraints\UrlValidator (@7968b5ae6a)
```
Comment (GrahamCampbell) [inline src/Illuminate/Validation/Validator.php]: is this hash symfony/symfony, or from the validator package?

**Comment by GrahamCampbell** [comment]
Thank you everyone. :)
